fix for Bug 969322
https://bugzilla.mozilla.org/show_bug.cgi?id=969322

:-moz-window-inactive triggers on window drag

This patch is needed to keep csd from showing inactive state on drag
whenever :-moz-window-inactive is used for styling.


--- a/widget/gtk/nsWindow.cpp	2019-10-16 20:30:30.000000000 +0000
+++ b/widget/gtk/nsWindow.cpp	2019-11-02 16:01:57.005586809 +0000
@@ -2951,6 +2951,14 @@
 
   DispatchDeactivateEvent();
 
+  if (mDrawInTitlebar) {
+    // DispatchDeactivateEvent() ultimately results in a call to
+    // nsGlobalWindowOuter::ActivateOrDeactivate(), which resets
+    // the mIsActive flag.  We call UpdateMozWindowActive() to keep
+    // the flag in sync with GDK_WINDOW_STATE_FOCUSED.
+    UpdateMozWindowActive();
+  }
+
   LOGFOCUS(("Done with container focus out [%p]\n", (void*)this));
 }
 
@@ -3174,6 +3182,9 @@
     mTitlebarBackdropState =
         !(aEvent->new_window_state & GDK_WINDOW_STATE_FOCUSED);
 
+    // keep mIsActive in sync with GDK_WINDOW_STATE_FOCUSED
+    UpdateMozWindowActive();
+
     ForceTitlebarRedraw();
   }
 
@@ -7028,6 +7039,19 @@
   return view->GetFrame();
 }
 
+void nsWindow::UpdateMozWindowActive() {
+  // Update activation state for the :-moz-window-inactive pseudoclass.
+  // Normally, this follows focus; we override it here to follow
+  // GDK_WINDOW_STATE_FOCUSED.
+  mozilla::dom::Document* document = GetDocument();
+  if (document) {
+    nsPIDOMWindowOuter* window = document->GetWindow();
+    if (window) {
+      window->SetActive(!mTitlebarBackdropState);
+    }
+  }
+}
+
 void nsWindow::ForceTitlebarRedraw(void) {
   MOZ_ASSERT(mDrawInTitlebar, "We should not redraw invisible titlebar.");
 
--- a/widget/gtk/nsWindow.h	2019-10-16 20:30:30.000000000 +0000
+++ b/widget/gtk/nsWindow.h	2019-11-02 16:01:59.275124120 +0000
@@ -608,6 +608,8 @@
 
   virtual int32_t RoundsWidgetCoordinatesTo() override;
 
+  void UpdateMozWindowActive();
+
   void ForceTitlebarRedraw();
 
   bool IsWaylandPopup();
